---
title: "Figure S3 & S4"
date: "`r format(Sys.time(), '%d %B, %Y')`"
format:
  html:
    embed-resources: true
    code-fold: show
params:
  fig.path: "`r paste0(params$fig.path)`" #./Figures/
editor_options: 
  chunk_output_type: console
---
## Heatmap of significant genes per cluster

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  fig.width     = 6.6929133858,
  fig.path      = params$fig.path,#"../Figures/",
  fig.align     = "center",
  message       = FALSE,
  warning       = FALSE,
  dev           = c("png"),
  dpi           = 300,
  fig.process = function(filename){
    new_filename <- stringr::str_remove(string = filename,
                                        pattern = "-1")
    fs::file_move(path = filename, new_path = new_filename)
    ifelse(fs::file_exists(new_filename), new_filename, filename)
  }
  )
#  setwd("~/work/Brolidens_work/Projects/Spatial_Microbiota/src/Manuscript")
```

```{r background_job, eval=FALSE, include=FALSE}
source("../../bin/render_with_jobs.R")

# quarto
# render_html_with_job(out_dir = lab_dir)
# fs::file_move(path = file, new_path = paste0(lab_dir, file))

# currently using quarto for github and kniter for html due to source code option 
render_git_with_job(fig_path = "./Figures/07/")
system2(command = "sed", stdout = TRUE,
        args = c("-i", "''","-e", 's/src=\\"\\./src=\\"\\.\\./g',
                 paste0("./md_files/", basename("./07_figures.md"))))

# kniter
knit_html_with_job(out_dir = "../lab_book/figure_07", fig_path = "./Figures/07/")
```

### Load data and libraries
```{r Load_data}
##################
# LOAD LIBRARIES #
##################
library(tidyverse)
library(Seurat)
library(SeuratObject)
library(tidyseurat)
library(cowplot)
library(ggrepel)
library(gridtext)
library(png)
library(grid)
#library(scatterpie)
library(patchwork)
library(openxlsx)
library(readxl)

# BiocManager::install("ComplexHeatmap")
library(ComplexHeatmap)

source("../../bin/spatial_visualization.R")
source("../../bin/plotting_functions.R")

#########
# PATHS #
#########
input_dir <- "../../results/06_DGE_condition_st_data/"
result_dir <- "../../results/09_figures/"
if( isFALSE(dir.exists(result_dir)) ) { dir.create(result_dir,recursive = TRUE) }
epi_clus <- "^5$|^6$|^7|^8" # res 0.75
# epi_clus <- "^11$|^6$|^7|^9" # res 1.0

ord <- c("Superficial", "Upper IM", "Lower IM", "Basal","1","4","0","3","2","9","10","11","12")
sample_id <- c("P020", "P045", "P050", "P057",
               "P008", "P031", "P080", "P044", "P026", "P105", 
               "P001", "P004", "P014", "P018", "P087", "P118",
               "P021", "P024", "P067", "P081", "P117" ) %>% set_names()

#############
# LOAD DATA #
#############
DATA <- readRDS(paste0("../../results/03_clustering_st_data/","seuratObj_clustered.RDS"))
pseudo_DEGs <- readRDS(paste0(input_dir,"Pseudobulk_across_DEGs.RDS"))
DEGs_table <- read_csv(paste0("../../results/06_DGE_condition_st_data/","DGEs_condition_wilcox.0.7.csv")) %>%
  mutate(layers = factor(.$layers, levels = ord))
#DATA <- readRDS(paste0(input_dir,"seuratObj_spatial_dist_SM.RDS"))

```

```{r stats}
# Find intersections for each layer
int.fun <- function(df){
  g_lists <- df %>% # g_lists <- DEGs_nest$data[[1]] %>%
    filter(p_val_adj < 0.05) %>%
    split(~comb) %>%
    map(., ~pull(.x, "gene"))
  
  int <- Reduce(intersect, g_lists )
  return(int)
}

DEGs_nest <- DEGs_table %>%
  filter(!(grepl("11|$^12$", .$layers))) %>%
  arrange(layers) %>%
  filter(p_val_adj < 0.05) %>% 
  nest(data = -c("comb", "layers")) %>%
  mutate(#UP = map_int(data, ~ nrow(filter(.x, p_val_adj <= 0.05 & avg_log2FC > 0))),
         #DOWN = map_int(data, ~ nrow(filter(.x, p_val_adj <= 0.05 & avg_log2FC < 0))),
         n = map_int(data, ~ nrow(.x)) ) %>%
  select(-data) %>%
  pivot_wider(., names_from = comb, values_from = n)

comp <- unique(DEGs_table$comb)
DEGs_nest <- DEGs_table %>%
  filter(!(grepl("11|$^12$", .$layers))) %>%
  arrange(layers) %>%
  #filter(p_val_adj < 0.05) %>% 
  filter(p_val < 0.001) %>% 
  #group_by(Regulation) %>%
  nest(data = -layers) %>%
  mutate(pval.0.001 = map_int(data, ~ nrow(filter(.x, p_val <= 0.001))),
         FDR.0.05 = map_int(data, ~ nrow(filter(.x, p_val_adj <= 0.05))),
         UP = map_int(data, ~ nrow(filter(.x, p_val_adj <= 0.05 & avg_log2FC > 0))),
         DOWN = map_int(data, ~ nrow(filter(.x, p_val_adj <= 0.05 & avg_log2FC < 0)))) %>%
  left_join(., DEGs_nest) #%>%
  #mutate(intersection = map(data, ~int.fun(.x)))

DEGs_nest


# Number of DEGs
WILK <- DEGs_table %>%
  filter(!(grepl("11|$^12$|^9", .$layers))) %>%
  filter(p_val_adj < 0.05) %>% .$gene %>% unique()

PAIR <- pseudo_DEGs %>%
  select(pairwise) %>% unnest(cols = c(pairwise)) %>% filter(FDR < 0.05) %>% .$Genes %>% unique()

ACK <- pseudo_DEGs %>%
  select(across) %>% unnest(cols = c(across)) %>% filter(FDR < 0.05) %>% .$symbol %>% unique()

setdiff(PAIR, WILK)
intersect(intersect(PAIR, WILK), ACK)

all_DEGs <- c(PAIR, WILK, ACK) %>% unique()

library(RColorBrewer)
myCol <- brewer.pal(3, "Pastel2")

# Chart
library(VennDiagram)
venn.diagram(
        x = list(WILK, PAIR, ACK),
        category.names = c("Wilcox" , "P-bulk paired" , "P-bulk across"),
        filename = '14_venn_diagramm.png',
        output=TRUE,
        
        # Output features
        imagetype="png" ,
        height = 480 , 
        width = 480 , 
        resolution = 300,
        compression = "lzw",
        
        # Circles
        lwd = 2,
        lty = 'blank',
        fill = myCol,
        
        # Numbers
        cex = .6,
        fontface = "bold",
        fontfamily = "sans",
        
        # Set names
        cat.cex = 0.6,
        cat.fontface = "bold",
        cat.default.pos = "outer",
        cat.pos = c(-27, 27, 135),
        cat.dist = c(0.055, 0.055, 0.085),
        cat.fontfamily = "sans",
        rotation = 1
)
```


### Volcano plots of differentially expressed genes per cluster
```{r 06a_DEGs_volcano_plot_epi, fig.width=19, fig.height=10, dev=c('png','pdf')}
# dev.new(width=19, height=7, noRStudioGD = TRUE) 
###########################
# VOLCANO PLOT EPITHELIUM #
###########################
DEGs_filt <- DEGs_table %>% 
  filter(grepl(epi_clus, .$subgroup)) %>%
  filter(p_val < 0.099) 

Volcano.fun_logFC(DEGs_filt, "layers", y.axis="p-value", 
                  lab_size = 4, dot_size = .7,
                  up=c(.6, 0.05), down = c(-.8, 0.05)) + # labeling: (logFC, p-value)
  ylab("avg. Log2 fold change") +
  facet_wrap(~comb, nrow = 1) +
  theme(plot.margin = unit(c(0, .5, .4, 1), "lines"),
        axis.text.x = element_text(vjust = .5, size = 14, angle = 30)) 

ggsave(paste0("./Figures/04/","volcano_epi.0.7.png"), width = 19, height = 10, bg = "white")
```

```{r 06b_DEGs_volcano_plot_subMuc, fig.width=19, fig.height=10, dev=c('png','pdf')}
# dev.new(width=25, height=10, noRStudioGD = TRUE) 
###########################
# VOLCANO PLOT SUBMUCOSA #
###########################
DEGs_filt <- DEGs_table %>% 
  filter(grepl("^1$|^4$|^0$|^3$|^2$|^10$", .$subgroup)) %>%
  #filter(!(grepl("L2-L4", .$comb))) %>%
  filter(p_val < 0.099) #%>%
  #filter(p_val_adj < 0.05)

Volcano.fun_logFC(DEGs_filt, "layers", y.axis="p-value", lab = T,
                  lab_size = 5, dot_size = .7,
                  up=c(1, 0.01), down = c(-1, 0.01)) +# labeling: (logFC, p-value)
  ylab("avg. Log2 fold change") +
  facet_wrap(~comb, nrow = 1) +
  theme(legend.position = "none", #y,x
        #axis.text.x = element_text(colour = col, margin = margin(t = 10)),
        #axis.ticks.x.bottom = element_line(colour = col),
        plot.margin = unit(c(0, .5, .4, 1), "lines"))

ggsave(paste0("./Figures/04/","volcano_sub.0.7.png"), width = 23, height = 10, bg = "white")


DEGs_filt <- pseudo_DEGs %>%
  mutate(across(c("pairwise", "across"), ~set_names(.x, .data[["layers"]]) )) %>%
  filter(grepl("^1$|^4$|^0$|^3$|^2$|^9|^10$", .$layers)) %>%
  pluck(., "pairwise") %>%
  bind_rows(., .id = "layers") %>%
  filter(PValue < 0.099)

Volcano.fun_logFC(DEGs_filt, "layers", y.axis="p-value", lab = F,
                  lab_size = 5, dot_size = .7,
                  up=c(.6, 0.05), down = c(-.6, 0.05)) +# labeling: (logFC, p-value)
  ylab("avg. Log2 fold change") +
  facet_wrap(~comparison, nrow = 1)
```

```{r barplot-function}
########################################
# OVERLAPING GENES BETWEEN SM AND DEGs #
########################################
# Get all combinations of layers
overlap_barplot.fun <- function(DEGs, fill = "Regulation", max_y = FALSE){
  # DEGs_list <- map(DEGs, ~set_names(.x$gene, .x$Regulation)) #map(DEGs, ~pull(.x, "gene"))
  # t <- c(DEGs_list) %>% 
  # tibble(genes=., layers=names(.))
  # 
  # DEGs_list <- rep(t$layers[1:11], 4)
  # SM <- c(rep("SM1", 11), rep("SM2", 11), rep("SM3", 11), rep("SM4", 11))
  # n <- paste0(SM , "_", DEGs_list)
  # 
  # # Find intersections for each combination
  # intersections <- function(df, comb) {
  #   genes_comb <- df$genes[df$layers %in% comb]
  #   Reduce(intersect, genes_comb)
  # }
  # int <- map2(DEGs_list, SM, ~intersections(t, c(.x, .y))) %>% set_names(., n)
  # 
  # # Combine combinations and intersections into a data frame
  # intersections_df <- tibble(name = names(int), intersections = int, SM = SM, layers = DEGs_list) %>%
  #   mutate(n = map_int(intersections, ~length(.x)))
  # 
  # # Print the intersections
  # knitr::kable(intersections_df, caption = "All overlaping genes for all combinations")
  #c("Pairwise","Across","Pairwise-Across","Wilcox","Wilcox-Across","Wilcox-Pairwise","Wilcox-Pairwise-Across")
  #col <- c("#80B1D3", "#FFFFB3", "#BEBADA","#8DD3C7", "#FB8072","#FDB462","#FCCDE5","#B3DE69")
  #col <- c("#E41A1C","#FF7F00","#C77CFF","#984EA3","#00BFC4","#00A9FF","#377EB8",
  #         "#CD9600","#7CAE00","#e0e067","#FFFF33","#FF61CC","#FF9DA7","#999999","#A65628")
  #col <- c("#569EFF","#FF706A")
  
  plot <- DEGs %>%
    ggplot(., aes( x=layers, fill=.data[[fill]])) +
      geom_bar(stat="count") + {if(isFALSE(max_y == FALSE)) ylim(0,max_y)} +
      scale_fill_manual(values = col) + theme_classic() +
    theme(axis.text.x = element_text(angle = 30, hjust = 1),
          legend.title = element_blank(),
          plot.margin = unit(c(0,0,0,.1), "inches"),
          legend.margin=margin(0,0,0,-5),
          legend.box.margin=margin(-0,0,-0,0),
          #legend.box.spacing = margin(0.5),
          axis.title = element_blank()) +
    facet_wrap(~comb, ncol = 6)
    #facet_grid(cols = vars(comb), rows = vars(test))
  
  return(tibble(plot=list(plot), genes= list(DEGs$gene)))
}
```

```{r DEG-module-overlap}
#######################
# GET DEGs TABLES #
######################
# pseudo
pseudo <- c("pairwise", "across") %>%
  set_names() %>%
  map(., ~pluck(pseudo_DEGs, .x)) %>%
  map(., ~bind_rows(.x, .id = "layers"))

pseudo$pairwise <- pseudo$pairwise %>% 
  as_tibble() %>%
  rename("comb"="comparison", "gene"="Genes") %>% #, "logFC_pairwise"="logFC", "FDR_pairwise"="FDR") 
  select(layers, comb, gene, logFC, FDR)

pseudo$across <- pseudo$across %>% 
  pivot_longer(cols = starts_with("logFC"), names_to = "comb", values_to = "logFC", names_prefix = "logFC.") %>%
  mutate(comb = paste0("L1-",.$comb)) %>%
  rename("gene"="symbol") %>%
  select(layers, comb, gene, logFC, FDR)

Wilcox <- DEGs_table %>%
  rename(FDR="p_val_adj", logFC="avg_log2FC")
  
DEGs <- bind_rows(list("pairwise"=pseudo[[1]], "across"=pseudo[[2]], "Wilcox"=Wilcox), .id = "test") %>%
  filter(!(grepl("11|^12$|^9", .$layers))) %>%
  mutate(layers = factor(.$layers, levels = ord)) %>%
  filter(FDR < 0.05) 

#######################
# GET GENE OVELAP #
######################
# remotes::install_github("vqf/nVennR")
library(nVennR)
col <- c("Wilcox","Wilcox-Pairwise","Wilcox-Across","Pairwise","Across","Pairwise-Across","Wilcox-Pairwise-Across")
d <- DEGs %>% 
  split(~test) %>% map(., ~unique(pluck(.x, "gene"))) %>%
  
  plotVenn(., 
         sNames = c("Wilcox", "Pairwise", "Across"),
         outFile=paste0("./Figures/03/", "Venn_SM1_L3xL4_sig.svg")
         )

overlap <- d$reg %>% tibble(col = col, gene = .) %>% unnest(gene)


DEGs <- DEGs  %>% 
  left_join(overlap, by = "gene")  %>%
  mutate(Regulation = ifelse(logFC > 0 & FDR < 0.05, "UP",
                            ifelse(logFC < 0 & FDR < 0.05,"DOWN", "NOT SIG.")))

DEG_genes <- DEG_list %>% map(., ~split(.x, ~layers)) 
module_genes <- GetModules(DATA) %>% split(~module) %>% map(., ~pull(.x, "gene_name"))

###########
# BARPLOT #
###########
col <- c("#80B1D3", "#FFFFB3", "#BEBADA","#8DD3C7", "#FB8072","#FDB462","#FCCDE5","#B3DE69")
a <- overlap_barplot.fun(DEGs, fill = "col")
col <- c("#569EFF","#FF706A")
b <- overlap_barplot.fun(DEGs, fill = "Regulation")
a$plot
p <- wrap_plots(c(a$plot, b$plot), ncol = 1) + plot_layout(axes = "collect")
ggsave(paste0("./Figures/04/","DEGs_barchart.png"), width = 14, height = 4)

pseudo_DEGs <- readRDS(paste0(input_dir,"Pseudobulk_across_DEGs.RDS"))

b <- set_names(pseudo_DEGs$pairwise, pseudo_DEGs$layers) %>%
  bind_rows(., .id = "Clusters") %>%
  filter(FDR < 0.05) %>%
   mutate(Regulation = ifelse(logFC > 0 & FDR < 0.05, "UP",
                            ifelse(logFC < 0 & FDR < 0.05,"DOWN", "NOT SIG."))) %>%
  rename(layers="Clusters", comb="comparison") %>%
  mutate(layers = factor(.$layers, levels=ord)) %>%
  overlap_barplot.fun(., max_y = 1150) 

p <- wrap_plots(c(a$plot, b$plot), ncol = 1)
ggsave(paste0("./Figures/04/","DEGs_barchart_pseudo.png"),p, width = 14, height = 4)

```

```{r top-tables, eval=FALSE}
top_aggregate.fun <- function(DEGs, n=40) {
  top <- DEGs %>%
    filter(p_val_adj <= 0.05) %>%
    mutate("Sig. in" = paste0(.$layers[cur_group_rows()], collapse = "|"), .by="gene") %>%
    {. ->> temp} %>%
    slice_max(n=n, order_by = tibble(abs(avg_log2FC), p_val), by="Regulation") %>%
    arrange(desc(avg_log2FC))
  
  # summary of regulation
  sig_count_uniqe <- temp %>% 
    dplyr::select(., gene, Regulation) %>% 
    unique() %>% group_by(Regulation) %>% 
    add_tally() %>%
    mutate(layers = "uniq total")
  
  sig_count <- temp %>% 
    dplyr::select(., gene, Regulation, layers) %>% 
    unique() %>% 
    group_by(layers, Regulation) %>% 
    add_tally(., ) %>% 
    bind_rows(sig_count_uniqe, .) %>%
    dplyr::select(., -gene) %>% 
    unique() %>% 
    pivot_wider( names_from = Regulation, values_from = n)
  
  g <- set_names(top$gene, top$Regulation) %>% .[!duplicated(.)]

  entz <- unlist(mget(org.Hs.egSYMBOL2EG, x=g, ifnotfound=NA))
  gene_name <- AnnotationDbi::select(org.Hs.eg.db, keys=entz, 
                                     columns=c("GENENAME"), keytype="ENTREZID") %>%
    bind_cols(gene = names(entz), .) 
  
  top <- top %>%
    filter(., gene %in% c(g[names(g)=="UP"][1:15], g[names(g)=="DOWN"][1:15])) %>%
    #bind_cols(., dplyr::select(gene_name,gene,"Gene name"="GENENAME")) %>%
    left_join(., dplyr::select(gene_name,gene,"Gene name"="GENENAME"), by="gene") %>%
    filter(n()==1 | n()>1 & p_val==max(p_val), .by="gene") %>% # filter duplicate genes 
    dplyr::select(gene, "Gene name", Cluster=layers, everything(), -c(1:2), 
                  -any_of(c("pct.1", "pct.2","cluster","log.pct.diff","morf")))
                  
  return(list("top15"=top, "sig DEGs"=sig_count))
}

library(org.Hs.eg.db)
top <- map_depth(DEGs_l,2, ~top_aggregate.fun(.x))

suppressWarnings({detach("package:org.Hs.eg.db", unload = TRUE)
                  detach("package:AnnotationDbi", unload = TRUE)})

# count of significant genes 
n <- c("Epithelial sig. genes", "Submucosal sig. genes", "E2 hig vs low sig. genes (epi)", "E2 hig vs low sig. genes (sub)")
top %>% 
  map_depth(., 2, ~pluck(.x, "sig DEGs")) %>% 
  flatten() %>% set_names(., n) %>% 
  imap(., ~add_row(.x, layers = c("", .y), .before = 1 )) %>% 
  bind_rows() %>%
  { . ->> tab_sig } %>%
  knitr::kable()

n <- c("Epithelium", "Submucosa", "E2_low_vs_high (epi)", "E2_low_vs_high (sub)", "Sum sig. DEGS", "DEGs condition", "DEGs E2 lvl")
top %>% 
  map_depth(., 2, ~pluck(.x, "top15")) %>% 
  flatten() %>% 
  append(., list(tab_sig, DEGs_table, DEGs_table_E2lvl)) %>% set_names(., n) %>%
  write.xlsx(., keepNA=TRUE, na.string="", overwrite=TRUE,
           file=paste0(result_dir, paste0("Supplemental Table 4.xlsx")))

# table with top 15 up and down DEGs
knitr::kable(top_epi$top15, caption = "Top 15 down and upregulated genes sorted by logFC")
```

```{r pseudobulk-bulk-correlation-plot}
###########################
# SPATIAL PSEUDOBULK DATA #
###########################
# get the count matrix for all cells
  sparse_mtrx <- DATA@assays$RNA@counts
# Compute pseudobulk
  mm <- Matrix::sparse.model.matrix(~0 + DATA$orig.ident)
  pseudobulk <- sparse_mtrx %*% mm
  colnames(pseudobulk) <- str_extract(colnames(pseudobulk), "P\\d\\d\\d")
  pseudobulk <- pseudobulk[, sample_id]
  
  # define the groups
  bulk.labels = groups
  
  dge.list <- DGEList(counts = pseudobulk)
  keep <- filterByExpr(dge.list)
  #dge.list <- dge.list[keep, , keep.lib.sizes = FALSE]
  
  dge.list <- calcNormFactors(dge.list)
  cpm <- cpm(dge.list, log=T) %>% as_tibble(., rownames = "symbol")

#############
# BULK DATA #
#############
trx_csv <- read_csv(paste0("/Users/vilkal/Raw_data/Bulk_Transcriptomics/Visit_3/Raw_counts_matrix_v3",".csv"))
sample_id <- c("P020","P045","P050","P057",
               "P008","P026","P031","P044","P080","P105", 
               "P001","P004","P014","P018","P087","P108","P118",
               "P021","P024","P067","P081","P117") %>% set_names()

bulk <- trx_csv %>% 
  select(symbol, all_of(sample_id)) %>%
  filter(rowSums(across(where(is.numeric)))!=0) %>%
  filter(!(is.na(.$symbol))) #%>%
  #filter(symbol %in% cpm$symbol)

# check for duplicated symbols
bulk %>% group_by(symbol) %>% summarize(n(), .groups='drop_last') %>% filter(`n()`>1)  %>% .$symbol

bulk_norm <- DGEList(bulk[,2:9], genes = bulk$symbol) %>%
  calcNormFactors(method = "TMM") %>%
  cpm(log=T) %>%
  as_tibble() %>%
  bind_cols(symbol = bulk$symbol, .) 

comp <- map(sample_id, ~full_join(select(cpm, symbol, ST_data=.x), select(bulk_norm, symbol, bulk_data=.x)) ) %>%
  bind_rows(., .id = "sample") %>%
  mutate(sample = factor(.$sample, levels = sample_id))

library(ggpubr) 
ggplot( comp, aes( y=ST_data, x=bulk_data ))+ 
  geom_point(pch = 1)+ 
  stat_cor(method = "pearson") + # , label.x = -0.5, label.y = 15
  geom_smooth(method=lm) + 
  #stat_regline_equation(color="blue", label.y = 13) +#geom_ +
  facet_wrap("sample", ncol = 4, strip.position = "top", shrink = F) +
  theme_light() + theme(text = element_text(size = 17)) 

```

```{r DEGs-heatmaps}
library(RColorBrewer)
heatmap_logFC_SM.fun <- function(mod, layers, test="across"){
  kME <- paste0("kME_", mod)
   modules <- GetModules(DATA)
  
  if(test == "across"){
    n <- 3
   d <- pseudo_DEGs %>%
    mutate(across = set_names(.data[["across"]], .data[["layers"]])) %>%
    pluck(., "across", layers) %>%
     left_join(., modules, by = c("symbol"="gene_name")) %>%
     filter(grepl(mod, .$module)) %>%
     filter(FDR < 0.05)
  }else{
    n <- 6
     d <- pseudo_DEGs %>%
      mutate(pairwise = set_names(.data[["pairwise"]], .data[["layers"]])) %>%
       pluck(., "pairwise", layers) %>%
       select(-c("PValue", "F", "logCPM")) %>%
       pivot_wider( names_from = comparison, values_from = c(logFC, FDR)) %>%
       rename_with(~ gsub("logFC_L1-", "logFC.", .x, fixed = TRUE)) %>%
       dplyr::rename(symbol = "Genes", FDR="FDR_L2-L4") %>%
       left_join(., modules, by = c("symbol"="gene_name")) %>%
       filter(grepl(mod, .$module)) %>%
       filter(FDR < 0.05)
   }
    
   if(nrow(d) != 0){
    #stopifnot(!(nrow(d) == 0))
     d <- d %>%
        mutate(diff = logFC.L4-logFC.L2) %>%
        arrange(desc(abs(`diff`))) %>%
        slice_head(n=100) %>%
        arrange(desc(logFC.L2)) %>%
        arrange(desc(logFC.L4)) %>%
        #arrange(desc( !!sym(kME))) %>%
        column_to_rownames(., var = "symbol")
  
    mat <- d %>% 
      na.omit(.) %>% .[1:n] %>%
      base::as.matrix(.) 
  
  # set min and max values for colour scale 
  mat[mat < -2] <- -2
  mat[mat > 2] <- 2
  
  h = nrow(mat) + max(str_width(colnames(mat)))

  paletteLength <- 100
  myColor <- colorRampPalette(rev(brewer.pal(n = 7, name ="RdBu")))(paletteLength)
  # myColor <- colorRampPalette(brewer.pal(n = 7, name ="RdBu"))(paletteLength)
  # length(breaks) == length(paletteLength) + 1
  # use floor and ceiling to deal with even/odd length pallettelengths
  myBreaks <- c(seq(min(mat), 0, length.out=ceiling(paletteLength/2) + 1), 
                seq(max(mat)/paletteLength, max(mat), length.out=floor(paletteLength/2)))

  
  p <- pheatmap::pheatmap(mat, 
           color=myColor, breaks=myBreaks,
           border_color = "transparent",
           legend_breaks = c(2, 1, 0, -1,-2),
           legend_labels = c("> 2", "1", "0", "-1", "<-2"),
           main = paste0(mod," ",layers), 
           height = unit(.1, "inches")*nrow(mat)+unit(.5, "inches"),
           width = unit(.1, "inches")*ncol(mat)+unit(2, "inches"),
           cluster_rows = F,
           treeheight_col = 0,
           annotation_row = select(d, "FDR"), #,"kME_SM1", "kME_SM2", "kME_SM3", "kME_SM4", "kME_SM5", "kME_SM6"
           silent = TRUE)$gtable
  
  return(tibble(p = list(p), h = h, genes = list(rownames(mat))))

   }
  }

mod <-  c("SM1", "SM2", "SM3", "SM4") %>% set_names()
l <- c("Superficial", "Upper IM", "Lower IM", "Basal","1","4","0","3","2","10") %>% set_names()

LogFC_heatmap.fun <- function(layers, test = "pairwise"){
  df <- map(mod, ~heatmap_logFC_SM.fun(.x, layers, test = test)) %>%
    bind_rows() #%>% #compact() %>% 
    #mutate(r_height = .$h/max(.$h))
  if(nrow(df) != 0){
    p <- plot_grid(plotlist = df$p, ncol=4) + theme(title = element_text(size=8 )) 
  return(tibble(p = list(p), h = max(df$h), genes = list(df$genes)) )}
}

plot_tbl <- map(l, ~LogFC_heatmap.fun(.x), test = "pairwise") %>%
  bind_rows(., .id = "id")

plot_tbl <- map(l, ~LogFC_heatmap.fun(.x, test = "across")) %>%
  bind_rows(., .id = "id") 

# individual files
pmap(plot_tbl, ~ggsave(paste0("./Figures/03/LogFC_pairwise/Module_gene_heatmap_",..1,".png"), 
                  ..2, width = 17, height = (.1*..3)+2, limitsize = F, bg = "white"))
pmap(plot_tbl, ~ggsave(paste0("./Figures/03/LogFC_across/Module_gene_heatmap_",..1,".png"), 
                  ..2, width = 17, height = (.1*..3)+2, limitsize = F, bg = "white"))

# multipage pdf:
ggsave(paste0("./Figures/03/LogFC_across/Module_gene_heatmap_","across",".pdf"), width = 15, height = 
       gridExtra::marrangeGrob(grobs=plot_tbl$p, ncol=1, nrow=1))
  
```


```{r violin}
  


violin.fun <- function(
    obj, 
    feature, 
    facet="feature", 
    group.by = "layers",
    fill="sample_name", 
    col_pal=NULL, 
    txt_size=7,
    n=1){
  if(is.null(col_pal)){col_pal <- c("#4E79A7","#F28E2B","#E15759","#76B7B2","#59A14F","#EDC948","#B07AA1","#FF9DA7","#9C755F","#BAB0AC") }
  
  obj <- obj %>%
    select(any_of(c(feature, facet, group.by, fill))) %>%
    pivot_longer(cols = any_of(feature), names_to = "feature", values_to = "val")
    
  m <- max(obj[["val"]])/n # try e.g 2
  obj %>%
    ggplot(aes(.data[[group.by]], .data[["val"]], fill=.data[[group.by]])) +
    geom_violin() + #ggtitle(feature) +
    geom_boxplot(fill = "transparent", notch = T) +
    geom_jitter(width = 0.3, alpha = 0.2, size=.1) +
    scale_fill_manual(values = col_pal) +
    my_theme + NoLegend() + ylim(c(0, m)) +
    theme(text = element_text(size = txt_size),
          axis.text.x = element_text(angle = 30, hjust=1),
          plot.title = element_text(hjust = 0.5),
          axis.title.y = element_blank(),
          axis.title.x = element_blank()) +
    facet_wrap(~.data[[facet]])
}

#################################
# IDENTIFY MARKER GENES IN DATA #
#################################
Cell_marker <- c("HLA") # "CD81", "SDC1"

rownames(DATA)[str_detect(rownames(DATA), paste0("^",Cell_marker,"$", collapse = "|"))]
f <- rownames(DATA)[str_detect(rownames(DATA), paste0("^",Cell_marker,"\\D", collapse = "|"))]
f <- rownames(DATA)[str_detect(rownames(DATA), paste0("^",Cell_marker,"\\d", collapse = "|"))]
rownames(DATA)[str_detect(rownames(DATA), paste0(Cell_marker, collapse = "|"))]

f <- "REV3L"
f <- c("HLA-DPB1", "HLA-DRA", "HLA-DRB1")
f <- c("LRP1","TIMP2","TIMP3","TIMP1","RECK")
DAT <- DATA %>%
  filter((grepl("^1$|^4$|^0$|^3$|^2$|^10$", .$Clusters))) %>%
  #filter((grepl("^5$|^6$|7|8", .$Clusters))) %>%
  mutate(., FetchData(., vars = c(f)) ) %>%
  #violin.fun(., feature=f,facet="layers", group.by = "groups") 
  violin.fun(., feature=f,facet="feature", group.by = "groups") 
  VlnPlot(., features = f, ncol = 6, group.by = "groups")

```

