---
title: "Supplemental Figure 5"
date: "`r format(Sys.time(), '%d %B, %Y')`"
format:
  html:
    embed-resources: true
    code-fold: show
params:
  fig.path: "`r paste0(params$fig.path)`" #./Figures/
editor_options: 
  chunk_output_type: console
---
## Heatmap of significant genes per cluster

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  fig.width     = 6.6929133858,
  fig.path      = params$fig.path,#"../Figures/",
  fig.align     = "center",
  message       = FALSE,
  warning       = FALSE,
  dev           = c("png"),
  dpi           = 300,
  fig.process = function(filename){
    new_filename <- stringr::str_remove(string = filename,
                                        pattern = "-1")
    fs::file_move(path = filename, new_path = new_filename)
    ifelse(fs::file_exists(new_filename), new_filename, filename)
  }
  )
#  setwd("~/work/Brolidens_work/Projects/Spatial_Microbiota/src/Manuscript")
```

```{r background_job, eval=FALSE, include=FALSE}
source("../../bin/render_with_jobs.R")

# quarto
# render_html_with_job(out_dir = lab_dir)
# fs::file_move(path = file, new_path = paste0(lab_dir, file))

# currently using quarto for github and kniter for html due to source code option 
render_git_with_job(fig_path = "./Figures/S3/")
system2(command = "sed", stdout = TRUE,
        args = c("-i", "''","-e", 's/src=\\"\\./src=\\"\\.\\./g',
                 paste0("./md_files/", basename("./Figure_S3.md"))))

# kniter
knit_html_with_job(out_dir = "../lab_book/figure_S3", fig_path = "./Figures/S3/")
```

### Load data and libraries
```{r Load_data}
##################
# LOAD LIBRARIES #
##################
library(tidyverse)
library(Seurat)
library(SeuratObject)
library(tidyseurat)
library(cowplot)
library(readxl)
library(vegan)
library(ggforce)

source("../../bin/spatial_visualization.R")
source("../../bin/plotting_functions.R")

#########
# PATHS #
#########
input_dir <- "../../results/08_spatial_dist/SM/"
input_dir <- "../../results/04_deconvolution_st_data/"

#############
# LOAD DATA #
#############
meta <- read_csv("../../data/Clinical_data/Spatial_clinical_data.csv")
# DATA <- readRDS(paste0("/Users/vilkal/work/Brolidens_work/Projects/Spatial_Microbiota/results/07_sp_microbes/","seuratObj_microbes_old.RDS"))

dataset_names <- c("ASV_Luminal_raw_counts", # Tissue, Boston run 1 (108 samples)
                   "ASV_Tissue_raw_counts")  # Tissue, Boston run 2+1 (93 sample)

datasets <- map(dataset_names, 
  ~read_excel(paste0("../../data/", "Suppl.Tbl.01 Abundance, diversity, BCs and ASV counts.xlsx"), sheet = .x)) %>% 
  set_names(., dataset_names)

sample_id <- c("P020", "P045", "P050", "P057",
               "P008", "P031", "P044", "P080", "P026", "P105", 
               "P001", "P004", "P014", "P018", "P087", "P118",
               "P021", "P024", "P067", "P081", "P117" ) %>% set_names()
```

```{r colour-pallet}
#####################################
# IDENTIFY MOST HIGLY ABUNDANT TAXA #
#####################################
# get top taxa by ordering max/median/mean percent tax in individual samples
# you can test what order you prefer by changing between max, median, and mean
d <- datasets[["ASV_Luminal_raw_counts"]] %>% 
  mutate(across(starts_with("P"), ~.x/sum(.x))) %>%
  rowwise() %>% mutate(., p = max(c_across(2:ncol(.)))) %>%
  select(1, p) %>% arrange(desc(p))

# gives list of taxa by decreasing percent
n <- d$Genus_taxa_luminal[1:24]

##################
# COLOUR PALETTE #
##################
n <- c('L. crispatus','L. acidophilus','L. crispatus/acidophilus','L. iners','L. other','L. jensenii','L. johnsonii',
       'L. gasseri/johnsonii/taiwanensis','L. reuteri', 'L. reuteri/oris/frumenti/antri',
       'Gardnerella','Prevotella','Atopobium','Sneathia','Megasphaera','Streptococcus',
       'Anaerococcus','Dialister','Mycoplasma','Bifidobacterium', 'Klebsiella', 'Citrobacter/Klebsiella',
       'Escherichia','Escherichia/Shigella')

cols <- c( "#A8EDFC","#A8EDFC","#A8EDFC","#87c7c0","#a9e7e4","#c2ebe2","#7fe2e9","#7fe2e9","#83dafb","#83dafb",# 
           "#be6a7d","#f1a6b1","#E3E6AD","#F8D0A4","#c4ce96","#9aacce","#e1caff","#abc5bf",
           "#ffffd4","#c0a2c1","#c8ffd5","#c8ffd5","#afb7ee","#ffc8d9","#ffc8d9","#e7b993","#c8ffd5",
           "#c4cea9","#a1b37d","#a6cca7","#d1b9ee","#88c29c",
           "#fdcc8a","#91c6f7","#f5f8bd","#8db1c5","#fab0aa","#7cb6b6","#96f3eb","#6ececc")

cols <- set_names(cols[1:length(n)], n)
cols <- c(cols, 'other'="gray90")
```

```{r functions}
# agglomerate all Lactobacillus species except for the ones specified by the vector (L) to "L. other"
agglom_L.fun <- function(df, taxa="taxa", format_W=T, counts="value"){
  # NB! specify weather the df is in long (format_W=FALSE) or wide format (format_W=TRUE)
  L <- c("crispatus", "iners","jensenii", "johnsonii", "reuteri", "acidophilus")
  L <- df[[taxa]][!(grepl( paste0(L, collapse = "|"), df[[taxa]]))]
  L <- gsub("(^Lactobacillus)(.+)", "L.\\2", L)
  L <- unique(L[grep("L\\.", L)])
  pattern <- paste0(L, collapse = "|")
  
  df <- df %>%
    mutate(!!(taxa) := gsub(pattern, "L. other", !!sym(taxa))) %>%
    {if(format_W) summarise(., across(where(is.numeric), sum), .by = taxa) else .} %>%
    {if(isFALSE(format_W)) summarise(., across(where(is.numeric), sum),  .by =-all_of(counts)) else .}
}

# order samples by percentage of specified taxa based on assigned microbiome group
# NB! I found that I got a cleaner looking graph by using gardnerella also for L4 (typ = identity) 
factor.fun <- function(df, type="stack"){
  l <- c("L1"="L. crispatus/acidophilus", "L2"="L. jensenii", "L3"="L. iners", "L4"="Gardnerella")
  if(type=="identity"){l <- c("L1"="L. crispatus/acidophilus", "L2"="L. iners", 
                              "L3"="Gardnerella", "L4"="Gardnerella")}
  imap(l, ~filter(df, gr==.y & taxa==.x) %>% 
               arrange(., desc(Percent)) %>% pull(., "ID") ) %>%
        unlist()}
```


```{r Taxa-bar-plot, fig.width=6, fig.height=6}
############
# FUNCTION #
############
# order individually by luminal groups
taxa_plot.fun <- function(type, area=F, txt_size=12){
  if(type=="identity"|type=="stack"){col <- "Percent"}else{col <- "value"}
  
  # create individual dfs for each group, in order to order taxa for each separately:
  d <- df %>% 
    {if(type=="identity") mutate(., ID = factor(ID, levels = factor.fun(., type="identity"))) else .} %>%
    split(~gr) %>% 
    map(., ~ .x %>%
          arrange(., desc(Percent)) %>%
          mutate(., taxa = factor(taxa, levels=unique(.$taxa))) ) 
  # check resulting taxa levels
  d %>% map(., ~levels(.x$taxa)) 
  
  ggplot() + 
    {if(area)list(
      geom_area(data = d$L1, aes(x=ID, y=.data[[col]], group=taxa, fill=taxa), position = type),
      geom_area(data = d$L2, aes(x=ID, y=.data[[col]], group=taxa, fill=taxa), position = type),
      geom_area(data = d$L3, aes(x=ID, y=.data[[col]], group=taxa, fill=taxa), position = type),
      geom_area(data = d$L4, aes(x=ID, y=.data[[col]], group=taxa, fill=taxa), position = type))}+
    {if(isFALSE(area))list(
      geom_col(data = d$L1, aes(x=ID, y=.data[[col]], group=taxa, fill=taxa), position = type),
      geom_col(data = d$L2, aes(x=ID, y=.data[[col]], group=taxa, fill=taxa), position = type),
      geom_col(data = d$L3, aes(x=ID, y=.data[[col]], group=taxa, fill=taxa), position = type),
      geom_col(data = d$L4, aes(x=ID, y=.data[[col]], group=taxa, fill=taxa), position = type))} +
    
    #{if(type=="identity") 
    #  geom_line(data = df, aes(x=ID, y=.data[[col]], group=taxa), color="white", size=.2)} + 
    scale_color_manual(values = cols, aesthetics = c("color", "fill") ) + theme_classic(base_size = txt_size) +
    scale_y_continuous(labels = scales::percent) + coord_cartesian(expand = F) +
    theme(axis.text.x = element_text(angle = 90, vjust = .5), 
          #axis.text = element_text(size = 10),
          #legend.text = 
          axis.title = element_blank(), legend.title = element_blank()) #legend.position=c(.9,.74),
}

######################
# TAXA AREA/BAR PLOT #
######################
# get taxa df
df <- datasets$ASV_Luminal_raw_counts %>% 
  pivot_longer(cols = -Genus_taxa_luminal, names_to = "ID") %>%
  filter(ID %in% sample_id) %>%
  left_join(., select(meta, ID, gr="Luminal_gr_v3"), by="ID") %>%
  # creates L. other category:
  agglom_L.fun(., "Genus_taxa_luminal", format_W=F) %>%
  # Non-abundant taxa to "other":
  mutate(taxa = ifelse(.$Genus_taxa_luminal %in% n, .$Genus_taxa_luminal, "other")) %>%
  mutate(taxa = factor(taxa, levels = c(n,"other"))) %>%
  mutate(Percent = value/sum(value), .by = "ID") %>%
  # Others.fun(., "taxa", percent = 0.001) %>%
  mutate(ID = factor(ID, levels = factor.fun(.))) ## order samples by percentage of taxa

# check if percentages add up to one
group_by(df, ID) %>% summarize(total_percent = sum(Percent)) 

d <- summarize(df, total_percent = sum(Percent), .by = c("taxa", "ID")) 

# get plot
# dev.new(width=6, height=6, noRStudioGD = TRUE) 
(A <- taxa_plot.fun(type = "fill", txt_size=10))

(A <- taxa_plot.fun(type = "fill", txt_size=10, area=T))
```

```{r NMDS, fig.width=6, fig.height=6}
#################
# PREPP DATSETS #
#################
samples <- sample_id
tax <- c(n, "other")
fix <- c("L. reuteri/oris/frumenti/antri", "L. crispatus/acidophilus", "L. gasseri/johnsonii/taiwanensis",
         "Citrobacter/Klebsiella", "Escherichia/Shigella", "Delftia/Roseateles", "Clostridium_IV","Clostridium_sensu_stricto", "Clostridium_XlVa")
new_n <- c("L. reuteri", "L. acidophilus", "L. johnsonii","Klebsiella", "Escherichia", "Delftia", "Clostridium", "Clostridium", "Clostridium") %>% set_names(., fix)

# uses the the 
datasets_ <- datasets %>%
  set_names(., c("L","T")) %>%
  map(., ~rename(., taxa = 1)) %>%
  # agglomerate all Lactobacillus species except for the ones specified by L to L. other
  map(., ~agglom_L.fun(.x, format_W = T) ) %>%
  # renames names with "/" in them to single species
  #map(., ~mutate(.x, taxa = ifelse( .$taxa %in% fix, new_n[.$taxa], .$taxa)) )  %>%
  # agglomerates any possible duplicates
  #map(., ~summarise(.x, across(where(is.numeric), sum), .by = "taxa") ) %>%
  
  map(., ~select(.x, taxa, any_of(samples)) ) %>%
  #map(., ~filter(.x, .$taxa %in% tax) ) %>%
  map(., ~column_to_rownames(.x, var = colnames(.x)[1]) ) %>%
  #imap(., ~.x %>% rename_with(function(x) c(paste(x, .y, sep = "_")) )) 
  map(., ~t(.x) ) %>%
  map(., ~as_tibble(.x, rownames="ID"))

##############
# META DATA #
##############
m <- meta %>%
  select(ID, gr_L="Luminal_gr_v3", gr_T="Tissue_gr_v3") %>%
  filter(ID %in% samples)

#############
# NMDS PLOT #
############# 
get_NMDS.fun <- function(df, gr="L"){
  # NMDS from https://github.com/giacomellolab/SpatialMetaTranscriptomics/blob/main/Compare_SmT_profiles.R
  example_NMDS=metaMDS(df, # Our raw counts samples (row) by species (col) matrix
                     k=2, trymax = 100, distance = "bray") # The number of reduced dimensions -> ,, autotransform = FALSE,distance ="euclidean"
  stressplot(example_NMDS)
  plot (example_NMDS)
  
  m <- m %>% arrange(paste0("gr_", gr))
  gr <- m[[paste0("gr_", gr)]]
  n <- set_names(gr, m$ID)
  
  #extract NMDS scores (x and y coordinates)
  data.scores = as.data.frame(example_NMDS$points)
  data.scores$gr =n[str_extract(row.names(data.scores), "P\\d\\d\\d")]
  data.scores$ID=row.names(data.scores)
  
  taxa.score = as.data.frame(example_NMDS$species)
  
  return(list("sample.score"= data.scores, "taxa.score"=taxa.score))
}

# single df
nmds_ls <- datasets_ %>% imap(., ~.x %>% column_to_rownames(., var = "ID") %>% get_NMDS.fun(., gr=.y))


#### Plot NMDS Single dataset at a time ####
B <- map(nmds_ls, ~ggplot(.x[[1]], 
                                 aes(x = MDS1, y = MDS2)) + 
  # geom_point(size = 4, aes( shape = Type, colour = Sample))+ 
    geom_point(size = 1, aes(fill = .data[["gr"]], colour = .data[["gr"]]), show.legend = F) +
    scale_fill_manual(values = c( "#56B4E9","#009E73","#CC79A7","#FC8D62", "gray51"),
                      aesthetics = c("colour","fill") ) +
    geom_mark_ellipse(aes(fill = .data[["gr"]], colour = .data[["gr"]]), expand = 0, show.legend = F) +
    geom_text(aes(label=ID), size=4, vjust = -.4, nudge_x = 0.02, color="gray51") +
    coord_equal(clip = 'off' ) +
    theme_classic(base_size = 10) +
    theme(legend.title=element_blank(),
          plot.margin = unit(c(0,0,-2,0), "lines"))
  )
# dev.new(width=6, height=6, noRStudioGD = TRUE) 
B$L
```

```{r FigureS3, fig.width=9, fig.height=5, dpi=1000}
#############################
# COMBINE ALL FIGURE PANELS #
#############################
# dev.new(width=9, height=5, noRStudioGD = TRUE)
FigureS3 <- plot_grid( A, B$L, ncol=2, rel_widths = c(1,.9)) 
# ggsave("./Figures/S3/Figure_S3.pdf", FigureS3, width = 9, height = 5, bg = "white", dpi = 1000)

FigureS3
```